domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
+
+        # probability is set to zero for wall modes
+        for n in nodes_list_help:
+            if isinstance(n, WallNode):
+                probabilities.append([0, n, 0])
+                zero_cells -= 1
+                nodes_list.remove(n)
+
+        # for regular nodes, half of probabilities should be zero, including wall nodes
+        if isinstance(node, RegularNode):
+            while zero_cells != 0:
+                random_node = rdm.choice(nodes_list)
+                nodes_list.remove(random_node)
+                probabilities.append([0, random_node, 0])
+                zero_cells -= 1
+
+        # number of nodes that can be reached with one action
+        # action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        if len(nodes_list) / len(ACTIONS) > 1.5:
+            action_len = math.ceil(len(nodes_list) / len(ACTIONS))
+        else:
+            action_len = math.floor(len(nodes_list) / len(ACTIONS))
+
+        # shuffle actions
+        random_actions = ACTIONS
+        rdm.shuffle(random_actions)
+        actions_copy = copy(random_actions)  # actions yet to be processed
 
-def create_graph(maze_env):
-    G = nx.DiGraph()
-    for cell, actions in maze_env.graph.items():
-        for action, target in actions.items():
-            source = (
-                cell.row,
-                cell.col,
-            )  # assuming each cell has row and col attributes
-            G.add_edge(source, target, action=action)
-    return G
+        for action in random_actions:
+
+            #  when it's the last action, all the left nodes are connected to the last action
+            if len(actions_copy) == 1:
+                nodes_for_current_action = nodes_list
+            else:
+                nodes_for_current_action = rdm.sample(nodes_list, action_len)
+                for n in nodes_for_current_action:
+                    nodes_list.remove(n)
+
+            # generates probabilities with sum of 1
+            non_zero_probabilities_action = self.generate_probabilities(nodes_for_current_action)
+
+            # appends (action, next_node, prob)
+            for i in range(len(nodes_for_current_action)):
+                probabilities.append([action, nodes_for_current_action[i], non_zero_probabilities_action[i]])
+
+            actions_copy.remove(action)  # removes processed action
+
+        return probabilities
+
+    @staticmethod
+    def generate_random_node(w, h):
+        prob = rdm.randint(1, 18)
+        if prob < 11:
+            return RegularNode(-1, w, h)
+        elif prob < 13:
+            return RegularNode(-10, w, h)
+        elif prob < 15:
+            return TerminalNode(-1, w, h)
+        elif prob < 17:
+            return WallNode(w, h)
+        else:
+            return TeleportNode(-2, w, h)
+
+    @staticmethod
+    def generate_probabilities(cells):
+        probabilities = np.random.rand(len(cells))
+        probabilities /= sum(probabilities)
+        return probabilities

```

Comment:
Kakvo je ovo cudno racunanje verovatnoce 😅

Reply:
239 - generise se lista duzine len(cells) sa random brojevima
240 - deli se svaki elemet te liste sa sumom svih, da bi ukupan zbir bio 1 ( u narodu poznato kao normalizacija )

Reply:
ahaaa to je lista ok ima smisla